### PR TITLE
Improvements to CLI usability and maintainability

### PR DESCRIPTION
**Pull request by fabianofranz**
Improves `kubectl` from an usability perspective by
1. Fixing how we handle terminal width in help. Some sections like the flags use the entire available width, while others like long descriptions breaks lines but don't follow a well established max width (screenshot below). This PR adds a new responsive writer that will adjust to terminal width and set 80, 100, or 120 columns as the max width, but not more than that given POSIX best practices and recommendations for better readability.
   terminal_width
2. Adds our own normalizers for long descriptions and cmd examples which allows us better control about how things like lists, paragraphs, line breaks, etc are printed. Features markdown support. Looks like `templates.LongDesc` and `templates.Examples` instead of `dedent.Dedend`.
3. Allows simple reordering and reuse of help and usage sections.
4. Adds `verify-cli-conventions.sh` which intends to run tests to make sure cmd developers are using what we propose as [kubectl conventions](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md). Just a couple simple tests for now but the framework is there and it's easy to extend.
5. Update [kubectl conventions](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md) to use our own normalizers instead of `dedent.Dedent`.

**Release note**:

``` release-note
Improves how 'kubectl' uses the terminal size when printing help and usage.
```

@kubernetes/kubectl

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34502)

### CODE REVIEW

**Comment by AdoHe** [comment]
@fabianofranz great job 👍

File: pkg/kubectl/cmd/templates/command_groups.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+			}
+		}
+	}
+	return false
+}
+
+func AddAdditionalCommands(g CommandGroups, message string, cmds []*cobra.Command) CommandGroups {
+	group := CommandGroup{Message: message}
+	for _, c := range cmds {
+		// Don't show commands that has no short description
```
Comment (kargakis) [inline pkg/kubectl/cmd/templates/command_groups.go]: s/has/have/

File: pkg/kubectl/cmd/annotate.go [hunk 2]
```diff
@@ -59,36 +59,36 @@ type AnnotateOptions struct {
...
 var (
-	annotate_long = dedent.Dedent(`
+	annotate_long = templates.LongDesc(`
 		Update the annotations on one or more resources.
 
-		An annotation is a key/value pair that can hold larger (compared to a label), and possibly not human-readable, data.
-		It is intended to store non-identifying auxiliary data, especially data manipulated by tools and system extensions.
-		If --overwrite is true, then existing annotations can be overwritten, otherwise attempting to overwrite an annotation will result in an error.
-		If --resource-version is specified, then updates will use this resource version, otherwise the existing resource-version will be used.
+		* An annotation is a key/value pair that can hold larger (compared to a label), and possibly not human-readable, data.
```
Comment (AdoHe) [inline pkg/kubectl/cmd/annotate.go]: why add `*` here?

File: hack/verify-cli-conventions.sh [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
+source "${KUBE_ROOT}/hack/lib/init.sh"
```
Comment (AdoHe) [inline hack/verify-cli-conventions.sh]: whether we should add this check to normal verify process? like check this when commit.

Comment (fabianofranz) [inline hack/verify-cli-conventions.sh]: Yep, it does it already. `verify` runs everything that satisfies `hack/verify-*.sh`, which includes this one.

https://github.com/fabianofranz/kubernetes/blob/001e58e83a82f881b4a753f33d08908d885aaa2e/hack/make-rules/verify.sh#L90

File: pkg/kubectl/cmd/annotate.go (hunk 2 above)
Comment (fabianofranz) [inline pkg/kubectl/cmd/annotate.go]: It's list items. Not strictly required, but if we don't use list items here we have to either give each one its own paragraph, or make it all part of a single paragraph (no line wrapping). Markdown unwraps single line breaks.

**Comment by fabianofranz** [comment]
What else? I'd like to get this in when possible given rebases here are painful. :)

File: pkg/kubectl/cmd/templates/markdown.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/russross/blackfriday"
+)
+
+const linebreak = "\n"
+
+// ASCIIRenderer is a blackfriday rendered intended for rendering markdown
```
Comment (kargakis) [inline pkg/kubectl/cmd/templates/markdown.go]: Can you add a
```go
var _ blackfriday.Renderer = ASCIIRenderer{}
```
to make it explicit that we are conforming to an external interface?

File: pkg/kubectl/cmd/templates/templates.go [hunk 5]
```diff
@@ -16,84 +16,75 @@ limitations under the License.
...
-	vars = `{{$isRootCmd := isRootCmd .}}` +
+	SectionVars = `{{$isRootCmd := isRootCmd .}}` +
 		`{{$rootCmd := rootCmd .}}` +
 		`{{$visibleFlags := visibleFlags (flagsNotIntersected .LocalFlags .PersistentFlags)}}` +
 		`{{$explicitlyExposedFlags := exposed .}}` +
 		`{{$optionsCmdFor := optionsCmdFor .}}` +
 		`{{$usageLine := usageLine .}}`
 
-	mainHelpTemplate = `{{with or .Long .Short }}{{. | trim}}{{end}}{{if or .Runnable .HasSubCommands}}{{.UsageString}}{{end}}`
+	SectionAliases = `{{if gt .Aliases 0}}Aliases:
```
Comment (kargakis) [inline pkg/kubectl/cmd/templates/templates.go]: Can you add godoc to these constants?

**Comment by kargakis** [comment]
Added a few more comments but looks good overall.

File: pkg/kubectl/cmd/templates/markdown.go (hunk 4 above)
Comment (fabianofranz) [inline pkg/kubectl/cmd/templates/markdown.go]: Done!

File: pkg/kubectl/cmd/templates/templates.go (hunk 5 above)
Comment (fabianofranz) [inline pkg/kubectl/cmd/templates/templates.go]: Godoc added.

File: pkg/util/term/term_writer_test.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+		"max 80":   {input: test, maxWidth: 80},
+		"max 120":  {input: test, maxWidth: 120},
+		"max 5000": {input: test, maxWidth: 5000},
+	}
+	for k, tc := range testcases {
+		b := bytes.NewBufferString("")
+		w := NewMaxWidthWriter(b, tc.maxWidth)
+		_, err := w.Write([]byte(tc.input))
+		if err != nil {
+			t.Errorf("%s: Unexpected error: '%s'", k, err)
```
Comment (fabianofranz) [inline pkg/util/term/term_writer_test.go]: Good catch, fixed.

**Comment by fabianofranz** [comment]
@kargakis all comments addressed, thanks!

**Comment by fabianofranz** [comment]
@janetkuo @AdoHe anything else?

**Comment by kargakis** [comment]
Tagging, additional comments can be addressed in follow-ups

**Pull request merged**